crates/codegen/runtime/cargo/src/runtime/diagnostic.rs

```diff
@@ -0,0 +1,65 @@
+use std::borrow::Cow;
+use std::error::Error;
+
+use crate::text_index::TextRange;
+
+#[repr(u8)]
+pub enum Severity {
+    Error = 1,
+    Warning = 2,
+    Information = 3,
+    Hint = 4,
+}
+
+pub trait Diagnostic: Error {
+    fn range(&self) -> TextRange;
+    fn code(&self) -> Option<Cow<'_, str>> {
```

Comment:
I'm using Cow here because in the common case, the code will be well-known and static, e.g. `TS1234`; it'd be a shame to always allocate here but I didn't want to limit the API and allow for Diagnostics to use something dynamic.

Reply:
Not sure I understand how `code()` will be used in Slang. Is this something we will surface from stack graphs?

Reply:
I can drop this from this patch if you'd like,however  I understood that we want to have the codes + glossary in the future. i.e. a quick code that allows lookup (`TS0123`, `CS0014`) + explainable entry in a user-visible database.

We don't need it for the parse errors, however, the more complex we get, the more need we'll have for it and it's a good idea to already bake it into the core trait here.

Reply:
should that database/glossary not be part of this system i.e. returning not a code but a documentation object of some form?

Reply:
My 2c, following our recent discussions, let's include only what we need for now, so that we don't confuse users with what this code is, or make any assumptions about how it is used. Once we decide to implement/ship a glossary, we can start defining these and adding it to our APIs.

Reply:
My 2c, following our recent discussions, I suggest including only what we need for now, so that we don't confuse users with what this code is, or make any assumptions about how it is used. Once we decide to implement/ship a glossary, we can start defining these and adding it to our APIs.

Reply:
Removed `code` for now.